engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -262,19 +272,54 @@ public Optional<Type> getType(String name) {
     return Optional.ofNullable(types.get(name));
   }
 
-  /** @return the raw method map held by this module */
-  public Map<Type, Map<String, Function>> getMethods() {
-    return methods;
+  /** @return a method for the given type */
+  public Function getMethodForType(Type tpe, String name) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    Map<String, Function> allTpeMethods = methods.get(tpeKey);
+    return allTpeMethods == null ? null : allTpeMethods.get(name);
+  }
+
+  /**
+   * Returns the names of methods for the given type.
+   *
+   * @param tpe the type in the scope
+   * @return names of methods
+   */
+  public Set<String> getMethodNamesForType(Type tpe) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    Map<String, Function> allTpeMethods = methods.get(tpeKey);
+    return allTpeMethods == null ? null : allTpeMethods.keySet();
+  }
+
+  /**
+   * Registers all methods of a type in the provided scope.
+   *
+   * @param tpe the methods of which type should be registered
+   * @param scope target scope where methods should be registered to
+   */
+  public void registerAllMethodsOfTypeToScope(Type tpe, ModuleScope scope) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    Map<String, Function> allTypeMethods = methods.get(tpeKey);
+    if (allTypeMethods != null) {
+      allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpeKey, name, fun));
+    }
+  }
+
+  /** @return methods for all registered types */
+  public List<Function> getAllMethods() {
```

Comment:
A nice trick to enforce immutability during compile time is `Collections<? extends Function>` - then the compiler refuses to call any `.add(obj)` except `.add(null)`.